### ISSUE TITLE
Failure to import mapping with code containing plus sign

### ISSUE THREAD

**Issue by bmamlin** [OpenConceptLab/ocl_issues#807]
When importing CEIL v2021-06-07, one mapping didn't make it into OCL:

[Sub-acute meningitis, presumed tuberculous (160441)](https://app.staging.openconceptlab.org/#/orgs/CIEL/sources/CIEL/concepts/160441/) mapped as narrower than `ICD-10` `A17.0+`:

```json
{
  "retired": false,
  "from_concept_url": "/orgs/CIEL/sources/CIEL/concepts/160441/",
  "type": "Mapping",
  "source": "CIEL",
  "to_source_url": "/orgs/WHO/sources/ICD-10-WHO/",
  "owner": "CIEL",
  "map_type": "NARROWER-THAN",
  "owner_type": "Organization",
  "external_id": "170409ABBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
  "to_concept_code": "A17.0+"
}
```

I'm guessing this mapping failed to import because the `to_concept_code` contains a character that would need to be URL encoded (i.e., the plus sign). Looking across all CIEL mappings, this is the only one that contains a plus sign in the `to_concept_code`.

While we still need to make sure any URLs created in the import scripting properly encode URLs, I wouldn't expect that these characters would need to be encoded within the JSON being imported. So, I'm assuming some part of the API's import process that's responsible for URL encoding failed in this case.

When I later tried to manually add the mapping to CIEL (on staging) through app.staging.openconceptlab.org (as `ocladmin`), it fails with the response:

> Parent, map_type, from_concept, to_source, to_concept_code must be unique.

which either means something might exist on the server for this mapping or the `to_concept_code` encoding failure is presenting with this error message.

**Comment by snyaggarwal** [comment]
@bmamlin This missing mapping does exist in OCL but its retired.
https://api.staging.openconceptlab.org/orgs/CIEL/sources/CIEL/mappings/?q=A17.0%2B&includeRetired=true

**Bugs:**
- This was showing as `A17.0 ` and not `A17.0+` in response and in UI  (space for +), I have fixed that part.
- It should have been updated if `update_if_exsists=True` was provided in the bulk import script but It would not have matched due to the `space` vs `+` issue in to_concept_code. This is also fixed now.

**Solution:**
Easiest is to just edit/unretire this mapping from UI.
From UI -- you should be able to search this by "A17.0+" and clicking on the "Include Retired" button.
Mapping Link -- https://app.staging.openconceptlab.org/#/orgs/CIEL/sources/CIEL/mappings/1131754/

**Comment by bmamlin** [comment]
Re-opening this issue, since there still seems to be a problem with plus signs:

* [CIEL "166403" &rarr; WHO-ICHI "PZX LC AH"](https://app.staging.openconceptlab.org/#/orgs/CIEL/sources/CIEL/mappings/1511210/) ([api](https://api.staging.openconceptlab.org/orgs/CIEL/sources/CIEL/mappings/1511210/))
* [CIEL "166405" &rarr; WHO-ICHI "PME ML AA"](https://app.staging.openconceptlab.org/#/orgs/CIEL/sources/CIEL/mappings/1511220) ([api](https://api.staging.openconceptlab.org/orgs/CIEL/sources/CIEL/mappings/1511220))

For both of these mappings, the `to_concept_code` is supposed to contain spaces; however, OCL shows them with plus signs. When edit their `to_concept_code` to replace the plus signs with spaces and update the mapping, the update appears to succeed, but the `to_concept_code` still contains plus signs.

@snyaggarwal,

Observed behavior:
1. Edit either of the above mappings, replace the plus signs with spaces in `to_concept_code`, and update
2. Mapping's `to_concept_code` still has plus signs instead of spaces

Expected behavior:
1. Edit either of the above mappings, replace the plug signs with spaces in `to_concept_code`, and update
2. Mapping's `to_concept_code` should have spaces instead of plus signs

**Comment by snyaggarwal** [comment]
@bmamlin This is fixed now on Staging

**Comment by bmamlin** [comment]
@snyaggarwal can we get this change promoted to Production? I'm trying to get CIEL v2021-06-07 released on Production (#810) and am running into these issues.

**Comment by bmamlin** [comment]
@snyaggarwal there still seems to be an issue on Staging with this:

For example, PIH has a mapping for [Cannabis-induced mental and behavior disorder (4683)](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/concepts/4683/) to PIH LiberiaMOH `"47 (Substance use disorder)"` [here](https://app.staging.openconceptlab.org/#/orgs/PIH/sources/PIH/mappings/1497656/). The `to_concept_code` returns in encoded form as `"47+%28Substance+use+disorder%29"`. I would expect the `to_concept_code` to be URL encoded _within URLs_; however, I would not expect the value itself to be encoded.

I would expect to see the decoded "47 (Substance use disorder)" instead of the encoded value in the UI:
![image](https://user-images.githubusercontent.com/860834/122793608-67472380-d289-11eb-8612-018e315c653c.png)

![image](https://user-images.githubusercontent.com/860834/122793425-323ad100-d289-11eb-8e7c-d310f8725052.png)

and I would expect this:

```bash
$ http -b "$OCL_API_URL/orgs/PIH/sources/PIH/mappings/1497656/" | jq .to_concept_code
"47+%28Substance+use+disorder%29"
```

to return `"47 (Substance use disorder)"` instead of the URL encoded value.

**Comment by snyaggarwal** [comment]
@bmamlin This specific mapping is fixed. The new mappings via UI/API or bulk import will be automatically fixed, the existing ones are the issue. The way to fix them now is to just edit it and correct the to_conept_code in the UI (remove special characters) and save it as an update. That will fix it

**Comment by bmamlin** [comment]
For the PIH import, I ended up copying the extra mappings from my diff file into sublime text and doing some quick search & replace to decoded the plus signs to spaces and percent codes to parentheses (quicker for me than scripting). I saved that as a json file and used `jq` to generate the needed commands. Something like:

```
jq -r "http --quiet PUT $OCL_API_URL\(.url) authentication:\"token $OCL_API_TOKEN\" \
  to_concept_code=\"\(.to_concept_code)\"" mappings-with-decoded-codes.json > update-mappings.sh
```

That created a shell script with all the `httpie` commands needed to patch the 250+ concepts. 🙂 Sorta hacky, but good enough for a one-time fix. I'll check another diff tonight after clinic and hopefully we'll have an exact match to what PIH sent us. 🤞  Thanks @snyaggarwal!
